### PR TITLE
Switch to VS17 and C++20

### PR DESCRIPTION
**Pull request by lhecker**
C++20 goodies include:
* 3-way comparison operator `<=>` and `operator==() = default`
* designated initializers
* constraints and concepts
* abbreviated function templates (`void foo(auto bar)`)
* pack-expansions in lambda init-captures
* heterogeneous lookups in `unordered_set` / `unordered_map`
* `consteval` / `constinit`
* `starts_with` / `ends_with` for `string` / `string_view`
* `midpoint`, `lerp`
* `<bit>`
* `<ranges>`
* `<span>`
* `<barrier>`
* `<latch>`
* `<semaphore>`

### CODE REVIEW

File: src/cascadia/TerminalConnection/CTerminalHandoff.cpp [hunk 1]
```diff
@@ -101,14 +106,16 @@ HRESULT CTerminalHandoff::EstablishPtyHandoff(HANDLE in, HANDLE out, HANDLE sign
 {
     try
     {
-        // Stash a local copy of _pfnHandoff before we stop listening.
+        std::unique_lock lock{ _mtx };
+
+        // s_StopListeningLocked sets _pfnHandoff to nullptr.
+        // localPfnHandoff is tested for nullness below.
+#pragma warning(suppress : 26429) // Symbol '...' is never tested for nullness, it can be marked as not_null (f.23).
         auto localPfnHandoff = _pfnHandoff;
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalConnection/CTerminalHandoff.cpp:114]: I am not the biggest fan of mechanical changes being mixed with functional and bugfix changes. :)

[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalConnection/CTerminalHandoff.cpp:114]: Like, I might like to backport a small race condition fix but I am not planning on backporting a whole toolchain change (for example!)

**Review by DHowett — CHANGES_REQUESTED** [review]
Blocking until 1.15 forks :)

File: .github/actions/spelling/patterns/patterns.txt [hunk 2]
```diff
@@ -1,11 +1,4 @@
-https://(?:(?:[-a-zA-Z0-9?&=]*\.|)microsoft\.com)/[-a-zA-Z0-9?&=_#\/.]*
-https://aka\.ms/[-a-zA-Z0-9?&=\/_]*
-https://www\.itscj\.ipsj\.or\.jp/iso-ir/[-0-9]+\.pdf
-https://www\.vt100\.net/docs/[-a-zA-Z0-9#_\/.]*
-https://www.w3.org/[-a-zA-Z0-9?&=\/_#]*
-https://(?:(?:www\.|)youtube\.com|youtu.be)/[-a-zA-Z0-9?&=]*
-https://(?:[a-z-]+\.|)github(?:usercontent|)\.com/[-a-zA-Z0-9?%&=_\/.+]*
-https://www.xfree86.org/[-a-zA-Z0-9?&=\/_#]*
+https?://\S+
```
[the remark this answers is not in the document]
Comment (skyline75489) [reply .github/actions/spelling/patterns/patterns.txt:1]: This is actually introduced by me, who tried to be more cautious than usual 😅

File: src/cascadia/TerminalConnection/CTerminalHandoff.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalConnection/CTerminalHandoff.cpp]: (Would you mind bringing this change in in a separate PR?)

**Review by zadjii-msft — APPROVED** [review]

File: src/cascadia/WinRTUtils/inc/Utils.h [hunk 3]
```diff
@@ -17,6 +17,9 @@ winrt::Windows::Foundation::IAsyncOperation<winrt::hstring> FilePicker(HWND pare
     DWORD flags{};
     THROW_IF_FAILED(fileDialog->GetOptions(&flags));
     THROW_IF_FAILED(fileDialog->SetOptions(flags | FOS_FORCEFILESYSTEM | FOS_NOCHANGEDIR | FOS_DONTADDTORECENT)); // filesystem objects only; no recent places
+
+    // The MSVC 14.31.31103 toolset seems to misdiagnose this line.
```
Comment (DHowett) [inline src/cascadia/WinRTUtils/inc/Utils.h:21]: 1: spell fix the word toolset

2: we should follow the STL team's convention and use "BODGY" to indicate compiler workarounds maybe?

**Review by DHowett — APPROVED** [review]

**Review by DHowett — APPROVED** [review]

**Pull request merged**